crossbeam-queue/src/spsc.rs

```diff
@@ -0,0 +1,449 @@
+    ///
+    /// let (p, c) = spsc::<i32>(100);
+    ///
+    /// assert_eq!(p.capacity(), 100);
+    /// ```
+    pub fn capacity(&self) -> usize {
+        self.inner.cap
+    }
+
+    /// Returns `true` if the queue is empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_queue::spsc;
+    ///
+    /// let (p, c) = spsc(100);
+    ///
+    /// assert!(p.is_empty());
+    /// p.push(1).unwrap();
+    /// assert!(!p.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns `true` if the queue is full.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_queue::spsc;
+    ///
+    /// let (p, c) = spsc(1);
+    ///
+    /// assert!(!p.is_full());
+    /// p.push(1).unwrap();
+    /// assert!(p.is_full());
+    /// ```
+    pub fn is_full(&self) -> bool {
+        self.len() == self.inner.cap
+    }
+
+    /// Returns the number of elements in the queue.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_queue::spsc;
+    ///
+    /// let (p, c) = spsc(100);
+    /// assert_eq!(p.len(), 0);
+    ///
+    /// p.push(10).unwrap();
+    /// assert_eq!(p.len(), 1);
+    ///
+    /// p.push(20).unwrap();
+    /// assert_eq!(p.len(), 2);
+    /// ```
+    pub fn len(&self) -> usize {
+        let head = self.inner.head.load(Ordering::Acquire);
+        let tail = self.tail.get();
+        self.head.set(head);
+        self.inner.distance(head, tail)
+    }
+}
+
+impl<T> fmt::Debug for Producer<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.pad("Producer { .. }")
+    }
+}
+
+/// The consumer side of a bounded single-producer single-consumer queue.
+///
+/// # Examples
+///
+/// ```
+/// use crossbeam_queue::{spsc, PopError};
+///
+/// let (p, c) = spsc(1);
+/// assert_eq!(p.push(10), Ok(()));
+///
+/// assert_eq!(c.pop(), Ok(10));
+/// assert_eq!(c.pop(), Err(PopError));
+///
+/// assert!(c.is_empty());
+/// assert!(!c.is_full());
+/// ```
+pub struct Consumer<T> {
+    /// The inner representation of the queue.
+    inner: Arc<Inner<T>>,
+
+    /// A copy of `inner.head` for quick access.
+    ///
+    /// This value is always in sync with `inner.head`.
+    head: Cell<usize>,
+
+    /// A copy of `inner.tail` for quick access.
+    ///
+    /// This value can be stale and sometimes needs to be resynchronized with `inner.tail`.
+    tail: Cell<usize>,
+}
+
+unsafe impl<T: Send> Send for Consumer<T> {}
+
+impl<T> Consumer<T> {
+    /// Attempts to pop an element from the queue.
+    ///
+    /// If the queue is empty, an error is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_queue::{spsc, PopError};
+    ///
+    /// let (p, c) = spsc(1);
+    /// assert_eq!(p.push(10), Ok(()));
+    ///
+    /// assert_eq!(c.pop(), Ok(10));
+    /// assert_eq!(c.pop(), Err(PopError));
+    /// ```
+    pub fn pop(&self) -> Result<T, PopError> {
```

Comment:
From a consumer point-of-view, why not at an `Option<T>`?

Reply:
Because `push` returns a `Result<(), PushError<T>>` so I'm returning a `Result` here for consistency. But otherwise I agree, `Option<T>` would be fine as well.